op-e2e/faultproofs/output_cannon_test.go

```diff
@@ -219,39 +219,55 @@ func TestOutputCannonDefendStep(t *testing.T) {
 	executor := uint64(1) // Different executor to the other tests to help balance things better
-	op_e2e.InitParallel(t, op_e2e.UsesCannon, op_e2e.UseExecutor(executor))
 
-	ctx := context.Background()
-	sys, l1Client := startFaultDisputeSystem(t)
-	t.Cleanup(sys.Close)
+	testPreimageStep := func(t *testing.T, precreatePreimage bool) {
+		op_e2e.InitParallel(t, op_e2e.UsesCannon, op_e2e.UseExecutor(executor))
 
-	disputeGameFactory := disputegame.NewFactoryHelper(t, ctx, sys)
-	game := disputeGameFactory.StartOutputCannonGame(ctx, "sequencer", 1, common.Hash{0x01, 0xaa})
-	require.NotNil(t, game)
-	outputRootClaim := game.DisputeLastBlock(ctx)
-	game.LogGameData(ctx)
+		ctx := context.Background()
+		sys, l1Client := startFaultDisputeSystem(t)
+		t.Cleanup(sys.Close)
 
-	game.StartChallenger(ctx, "sequencer", "Challenger", challenger.WithPrivKey(sys.Cfg.Secrets.Alice))
+		disputeGameFactory := disputegame.NewFactoryHelper(t, ctx, sys)
+		game := disputeGameFactory.StartOutputCannonGame(ctx, "sequencer", 1, common.Hash{0x01, 0xaa})
+		require.NotNil(t, game)
+		outputRootClaim := game.DisputeLastBlock(ctx)
+		game.LogGameData(ctx)
 
-	// Wait for the honest challenger to dispute the outputRootClaim. This creates a root of an execution game that we challenge by coercing
-	// a step at a preimage trace index.
-	outputRootClaim = outputRootClaim.WaitForCounterClaim(ctx)
-	provider := game.CreateCannonTraceProvider(ctx, "sequencer", outputRootClaim, challenger.WithPrivKey(sys.Cfg.Secrets.Alice))
-	preimageTraceIndex, err := provider.FindStepReferencingPreimage(ctx, 0)
-	require.NoError(t, err)
-	game.ChallengeIntoPosition(ctx, provider, outputRootClaim, preimageTraceIndex)
+		game.StartChallenger(ctx, "sequencer", "Challenger", challenger.WithPrivKey(sys.Cfg.Secrets.Alice))
+
+		// Wait for the honest challenger to dispute the outputRootClaim. This creates a root of an execution game that we challenge by coercing
+		// a step at a preimage trace index.
+		outputRootClaim = outputRootClaim.WaitForCounterClaim(ctx)
+		provider := game.CreateCannonTraceProvider(ctx, "sequencer", outputRootClaim, challenger.WithPrivKey(sys.Cfg.Secrets.Alice))
+		preimageTraceIndex, err := provider.FindStepReferencingPreimage(ctx, 0)
+		require.NoError(t, err)
+
+		execDepth := game.ExecDepth(ctx)
+		_, _, preimageData, err := provider.GetStepData(ctx, types.NewPosition(execDepth, big.NewInt(int64(preimageTraceIndex))))
+		require.NoError(t, err)
+		if precreatePreimage {
+			game.UploadPreimage(ctx, preimageData, sys.Cfg.Secrets.Alice)
+			require.True(t, game.PreimageExistsInOracle(ctx, preimageData))
+			game.ChallengeIntoPosition(ctx, provider, outputRootClaim, preimageTraceIndex)
+		} else {
+			game.ChallengeIntoPosition(ctx, provider, outputRootClaim, preimageTraceIndex)
+			// Ensure that the honest challenger uploaded the preimage for the correct index
+			require.True(t, game.PreimageExistsInOracle(ctx, preimageData))
+		}
```

Comment:
I think the parent PR will push most of this into a helper function so would be good to just give it an option to pre-upload the preimage for us so that's in the helper too.